packages/go_router/lib/src/match.dart

```diff
@@ -149,3 +146,281 @@ class ImperativeRouteMatch<T> extends RouteMatch {
+
+    // Also pop ShellRoutes when there are no subsequent route matches
+    while (newMatches.isNotEmpty && newMatches.last.route is ShellRouteBase) {
+      newMatches.removeLast();
+    }
+    // Removing ImperativeRouteMatch should not change uri and pathParameters.
+    if (match is ImperativeRouteMatch) {
+      return _copyWith(matches: newMatches);
+    }
+
+    final String fullPath = _generateFullPath(
+        newMatches.where((RouteMatch match) => match is! ImperativeRouteMatch));
+    // Need to remove path parameters that are no longer in the fullPath.
+    final List<String> newParameters = <String>[];
+    patternToRegExp(fullPath, newParameters);
+    final Set<String> validParameters = newParameters.toSet();
+    final Map<String, String> newPathParameters =
+        Map<String, String>.fromEntries(
+      pathParameters.entries.where((MapEntry<String, String> value) =>
+          validParameters.contains(value.key)),
+    );
+    final Uri newUri =
+        uri.replace(path: patternToPath(fullPath, newPathParameters));
+    return _copyWith(
+      matches: newMatches,
+      uri: newUri,
+      pathParameters: newPathParameters,
+    );
+  }
+
+  /// An optional object provided by the app during navigation.
+  Object? get extra => matches.isEmpty ? null : matches.last.extra;
+
+  /// The last matching route.
+  RouteMatch get last => matches.last;
+
+  /// Returns true if the current match intends to display an error screen.
+  bool get isError => error != null;
+
+  /// Returns the error that this match intends to display.
+  Exception? get error => matches.firstOrNull?.error;
+
+  /// The routes for each of the matches.
+  List<RouteBase> get routes => matches.map((RouteMatch e) => e.route).toList();
+
+  RouteMatchList _copyWith({
+    List<RouteMatch>? matches,
+    Uri? uri,
+    Map<String, String>? pathParameters,
+  }) {
+    return RouteMatchList(
+        matches: matches ?? this.matches,
+        uri: uri ?? this.uri,
+        pathParameters: pathParameters ?? this.pathParameters);
+  }
+
+  @override
+  bool operator ==(Object other) {
+    if (other.runtimeType != runtimeType) {
+      return false;
+    }
+    return other is RouteMatchList &&
+        const ListEquality<RouteMatch>().equals(matches, other.matches) &&
+        uri == other.uri &&
+        const MapEquality<String, String>()
+            .equals(pathParameters, other.pathParameters);
+  }
+
+  @override
+  int get hashCode {
+    return Object.hash(
+      Object.hashAll(matches),
+      uri,
+      Object.hashAllUnordered(
+        pathParameters.entries.map<int>((MapEntry<String, String> entry) =>
+            Object.hash(entry.key, entry.value)),
+      ),
+    );
+  }
+
+  @override
+  String toString() {
+    return '${objectRuntimeType(this, 'RouteMatchList')}($fullPath)';
+  }
+}
+
+/// Handles encoding and decoding of [RouteMatchList] objects to a format
+/// suitable for using with [StandardMessageCodec].
```

Comment:
Is there any benefit / reason this doesn't extend the [Codec](https://api.dart.dev/stable/3.0.3/dart-convert/Codec-class.html) class from dart:convert?

Reply:
fixed